### PR TITLE
Restore blocks from temp_block table - Closes #3775

### LINKED ISSUE #3775: Blocks need to be restored from temporary table upon failure or application restart
### Expected behavior
Whenever the node crashes or upon failure during one of the sync processes we roll back to the original state by applying the blocks in the temp table and continue the operating of the node.

Functions to be implemented:
- `restoreBlocks()`

1. Read blocks from temp table
2. Apply blocks (blocks verify process - do not care about how it is implemented - just call this)
3. Remove temp blocks after the apply process is successful

### Design For Issue
Place the `restoreBlocks` in `blocks/blocks`.
The synchronizer class should injects `blocks module` as a dependency so we can access `restoreBlocks` inside.

### Which version(s) does this affect? (Environment, OS, etc...)
`feature/improve-bft`

### Tasks list
- [ ] Add functionality for applying block from blocks temporary table.
- [ ] Restore blocks from the temporary table when the node restarts.
- [ ] Restore blocks from the temporary table when the synchronization fails.
- [ ] Write integration tests.

### PR DESCRIPTION
**Pull request by michielmulders**
### What was the problem?
No functionality for restoring and re-applying blocks.

### How did I solve it?
`restoreBlocks` util function that can be used inside of BSM or FSM in case we want to restore the original chain.

**Why not implement in synchronizer class?**
The original idea was to work with a base class (-> synchronizer) that holds the common logic for both sync mechanisms. However, now the synchronizer class has been converted in some orchestrator for registering both sync mechanisms and running them, we can not place `restoreBlocks` functionality inside it (creates a circular dependency).

Therefore, a utils file has been created that holds only one util function. We expect more util functions that hold common logic between both sync mechanisms. The disadvantage of this approach currently, we have to inject blocks and processor module to `restoreBlocks`.

### How to manually test it?
Will be implemented with one of the fork choices where the chain has to be stored.
Covered with unit tests.

### Review checklist

- [x] The PR resolves #3775 
- [x] All new code is covered with unit tests
- [ ] Relevant integration / functional tests are added
- [x] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [x] Documentation has been added/updated

### CODE REVIEW

File: framework/src/modules/chain/processor/processor.js [hunk 1]
```diff
@@ -184,7 +184,7 @@ class Processor {
 	}
 
 	// processValidated processes a block assuming that statically it's valid
-	async processValidated(block) {
+	async processValidated(block, removeFromTempTable = false) {
```
Comment (limiaspasdaniel) [inline framework/src/modules/chain/processor/processor.js]: Can we include this property in an object literal? This way we can name the parameters.

Comment (michielmulders) [reply framework/src/modules/chain/processor/processor.js]: Yes will do, thanks!

File: framework/src/modules/chain/synchronizer/utils.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+const restoreBlocks = async (blocksModule, processorModule, tx) => {
+	const tempBlocks = await blocksModule.getTempBlocks(tx);
+
+	if (tempBlocks.length === 0) {
+		throw new Error('Temp_block table is empty');
+	}
+
+	for (const block of tempBlocks) {
+		await processorModule.processValidated(block, true);
```
Comment (limiaspasdaniel) [inline framework/src/modules/chain/synchronizer/utils.js]: Need to update this for the new argument :)

File: framework/src/modules/chain/synchronizer/utils.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * 3. Each block gets deleted from temp_block table when its being applied
+ *
+ * @param {Object} tx - database transaction
+ * @return {Promise<Boolean>|Error} - returns true when successfully restoring blocks
+ */
+const restoreBlocks = async (blocksModule, processorModule, tx) => {
+	const tempBlocks = await blocksModule.getTempBlocks(tx);
+
+	if (tempBlocks.length === 0) {
+		throw new Error('Temp_block table is empty');
```
Comment (shuse2) [inline framework/src/modules/chain/synchronizer/utils.js]: do we need to throw error in the case it doesn't exist?
Maybe we can just return false and end the execution?

File: framework/test/jest/unit/specs/modules/chain/synchronizer/utils.spec.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+			// Arrange
+			const blocks = [{ id: 'block1' }, { id: 'block2' }];
+			blocksMock.getTempBlocks = jest.fn().mockReturnValue(blocks);
+
+			// Act
+			const result = await restoreBlocks(blocksMock, processorMock, stubs.tx);
+
+			// Assert
+			expect(blocksMock.getTempBlocks).toHaveBeenCalledWith(stubs.tx);
+			expect(processorMock.processValidated).toHaveBeenCalledTimes(2);
```
Comment (shuse2) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/utils.spec.js]: probably these cases are worth to have another `it` phrase to describe?

Comment (michielmulders) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/utils.spec.js]: Agree, makes sense, thanks!

**Pull request merged**